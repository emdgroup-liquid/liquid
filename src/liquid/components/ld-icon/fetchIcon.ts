import { getAssetPath } from '@stencil/core'

const iconCache = {}
const requestCache = {}

export async function fetchIcon(icon: string): Promise<string> {
  if (iconCache[icon]) {
    return iconCache[icon]
  }
  if (!requestCache[icon]) {
    requestCache[icon] = fetch(getAssetPath(`./assets/${icon}.svg`))
      .then((resp) => resp.text())
      .catch((err) => {
        console.error(`"${icon}" is not a valid name`, err)
        return ''
      })
  }

  const path = await requestCache[icon]
  iconCache[icon] = path

  return path
}
