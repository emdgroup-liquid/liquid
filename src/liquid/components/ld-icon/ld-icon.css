:host {
  display: inline-flex;

  > div {
    display: inline-flex;
  }
}
