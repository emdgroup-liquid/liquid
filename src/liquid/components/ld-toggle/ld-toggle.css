.ld-toggle {
  --ld-toggle-height: 2rem;
  --ld-toggle-width: 3.375rem;
  --ld-toggle-inner-space: 0.125rem;

  align-items: center;
  display: flex;
  height: var(--ld-toggle-height);
  min-width: auto !important;
  position: relative;
  /* animations triggered by user interactions on single component instances are probably not a performance issue */
  /* stylelint-disable-next-line plugin/no-low-performance-animation-properties */
  transition: background-color 200ms ease-in-out;
  width: var(--ld-toggle-width);

  input {
    background-color: var(--ld-col-rblck4);
    border-radius: var(--ld-br-full);
    appearance: none;
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: 0;

    &:not(:disabled):not([aria-disabled='true']) {
      cursor: pointer;
    }

    &:checked {
      ~ .ld-toggle__knob {
        transform: translateX(
          calc(var(--ld-toggle-width) - var(--ld-toggle-height))
        );
      }

      ~ .ld-toggle__icon-start {
        color: var(--ld-col-rblck6);
      }

      ~ .ld-toggle__icon-end {
        color: var(--ld-col-wht);
      }
    }

    &:disabled,
    &[aria-disabled='true'] {
      background-color: var(--ld-col-bg-g);

      &:checked ~ .ld-toggle__icon-start {
        color: var(--ld-col-rblck4);
      }

      &:not(:checked) ~ .ld-toggle__icon-end {
        color: var(--ld-col-rblck4);
      }
    }
  }

  .ld-toggle__knob {
    background-color: var(--ld-col-wht);
    border-radius: var(--ld-br-full);
    display: block;
    height: calc(var(--ld-toggle-height) - var(--ld-toggle-inner-space) * 2);
    margin: var(--ld-toggle-inner-space);
    pointer-events: none;
    transition: transform 200ms ease-in-out;
    width: calc(var(--ld-toggle-height) - var(--ld-toggle-inner-space) * 2);
    z-index: 1;
  }
}

.ld-toggle--large {
  --ld-toggle-height: 2.5rem;
  --ld-toggle-width: 4.1875rem;
  --ld-toggle-inner-space: 0.1875rem;

  &.ld-toggle--with-icons {
    --ld-toggle-width: 6rem;

    .ld-toggle__icon-start,
    .ld-toggle__icon-end {
      margin: 0.75rem;
    }
  }
}

.ld-toggle--with-icons {
  --ld-toggle-width: 4.5rem;

  input {
    background-color: var(--ld-col-rblck1) !important;

    &:invalid ~ .ld-toggle__knob {
      background-color: var(--ld-col-rr-default) !important;
    }

    &:disabled,
    &[aria-disabled='true'] {
      background-color: var(--ld-col-bg-g) !important;

      ~ .ld-toggle__knob {
        background-color: var(--ld-col-wht);
      }
    }
  }

  .ld-toggle__icon-start,
  .ld-toggle__icon-end {
    margin: 0.5rem;
  }
}

.ld-toggle input:invalid {
  background-color: var(--ld-col-rr-default);
}

/* .ld-theme-ocean -> default */
.ld-toggle,
.ld-theme-ocean .ld-toggle,
[class*='ld-theme'] .ld-theme-ocean .ld-toggle {
  input {
    &:checked {
      background-color: var(--ld-thm-ocean-bg-primary);

      &:disabled,
      &[aria-disabled='true'] {
        background-color: var(--ld-col-rb2);
      }
    }
  }

  &--with-icons .ld-toggle__knob {
    background-color: var(--ld-thm-ocean-bg-primary);
  }
}

.ld-toggle__icon-start,
.ld-toggle__icon-end {
  display: flex;
  align-items: center;
  pointer-events: none;
  position: absolute;
  /* animations triggered by user interactions on single component instances are probably not a performance issue */
  /* stylelint-disable-next-line plugin/no-low-performance-animation-properties */
  transition: color 200ms ease-in-out;
  z-index: 1;

  &:empty {
    display: none;
  }
}

.ld-toggle__icon-start {
  color: var(--ld-col-wht);
  left: 0;
}

.ld-toggle__icon-end {
  color: var(--ld-col-rblck6);
  right: 0;
}

.ld-theme-tea,
[class*='ld-theme'] .ld-theme-tea {
  .ld-toggle {
    input {
      &:checked {
        background-color: var(--ld-thm-tea-bg-primary);

        &:disabled,
        &[aria-disabled='true'] {
          background-color: var(--ld-col-rg2);
        }
      }
    }

    &--with-icons .ld-toggle__knob {
      background-color: var(--ld-thm-tea-bg-primary);
    }
  }
}

.ld-theme-bubblegum,
[class*='ld-theme'] .ld-theme-bubblegum {
  .ld-toggle {
    input {
      &:checked {
        background-color: var(--ld-thm-bubblegum-bg-primary);

        &:disabled,
        &[aria-disabled='true'] {
          background-color: var(--ld-col-rp2);
        }
      }
    }

    &--with-icons .ld-toggle__knob {
      background-color: var(--ld-thm-bubblegum-bg-primary);
    }
  }
}

.ld-theme-shake,
[class*='ld-theme'] .ld-theme-shake {
  .ld-toggle {
    input {
      &:checked {
        background-color: var(--ld-thm-shake-bg-primary);

        &:disabled,
        &[aria-disabled='true'] {
          background-color: var(--ld-col-rp2);
        }
      }
    }

    &--with-icons .ld-toggle__knob {
      background-color: var(--ld-thm-shake-bg-primary);
    }
  }
}

.ld-theme-solvent,
[class*='ld-theme'] .ld-theme-solvent {
  .ld-toggle {
    input {
      &:checked {
        background-color: var(--ld-thm-solvent-bg-primary);

        &:disabled,
        &[aria-disabled='true'] {
          background-color: var(--ld-col-rp2);
        }
      }
    }

    &--with-icons .ld-toggle__knob {
      background-color: var(--ld-thm-solvent-bg-primary);
    }
  }
}
