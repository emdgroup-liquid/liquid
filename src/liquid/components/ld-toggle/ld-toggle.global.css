.ld-theme-ocean,
[class*='ld-theme'] .ld-theme-ocean {
  .ld-toggle {
    --ld-toggle-bg-color-checked: var(--ld-thm-ocean-bg-primary);
    --ld-toggle-bg-color-checked-disabled: var(--ld-col-rb2);
    --ld-toggle-with-icons-knob-color: var(--ld-thm-ocean-bg-primary);
  }
}

.ld-theme-tea,
[class*='ld-theme'] .ld-theme-tea {
  .ld-toggle {
    --ld-toggle-bg-color-checked: var(--ld-thm-tea-bg-primary);
    --ld-toggle-bg-color-checked-disabled: var(--ld-col-rg2);
    --ld-toggle-with-icons-knob-color: var(--ld-thm-tea-bg-primary);
  }
}

.ld-theme-bubblegum,
[class*='ld-theme'] .ld-theme-bubblegum {
  .ld-toggle {
    --ld-toggle-bg-color-checked: var(--ld-thm-bubblegum-bg-primary);
    --ld-toggle-bg-color-checked-disabled: var(--ld-col-rp2);
    --ld-toggle-with-icons-knob-color: var(--ld-thm-bubblegum-bg-primary);
  }
}

.ld-theme-shake,
[class*='ld-theme'] .ld-theme-shake {
  .ld-toggle {
    --ld-toggle-bg-color-checked: var(--ld-thm-shake-bg-primary);
    --ld-toggle-bg-color-checked-disabled: var(--ld-col-rp2);
    --ld-toggle-with-icons-knob-color: var(--ld-thm-shake-bg-primary);
  }
}

.ld-theme-solvent,
[class*='ld-theme'] .ld-theme-solvent {
  .ld-toggle {
    --ld-toggle-bg-color-checked: var(--ld-thm-solvent-bg-primary);
    --ld-toggle-bg-color-checked-disabled: var(--ld-col-rp2);
    --ld-toggle-with-icons-knob-color: var(--ld-thm-solvent-bg-primary);
  }
}
