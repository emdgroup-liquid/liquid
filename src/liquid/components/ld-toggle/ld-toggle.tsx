import { Component, Element, h, Host, Prop } from '@stencil/core'
import '../../components' // type definitions for type checks and intelliSense
import { getClassNames } from '../../utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-toggle',
  styleUrl: 'ld-toggle.css',
  shadow: false,
})
export class LdToggle {
  @Element() element: HTMLElement
  private hasIcons: boolean

  /** Size of the toggle. */
  @Prop() size?: 'sm' | 'lg'

  /** Disabled state of the toggle. */
  @Prop() disabled = false

  /** Alternative disabled state that keeps element focusable */
  @Prop() ariaDisabled: string

  /** The input value. */
  @Prop({ mutable: true, reflect: true }) checked: boolean

  /** Set this property to `true` in order to mark the toggle as required. */
  @Prop() required: false

  componentWillLoad() {
    this.hasIcons =
      !!this.element.querySelector('[slot="icon-start"]') ||
      !!this.element.querySelector('[slot="icon-end"]')
  }

  private handleBlur(event) {
    setTimeout(() => {
      this.element.dispatchEvent(event)
    })
  }

  private handleFocus(event) {
    setTimeout(() => {
      this.element.dispatchEvent(event)
    })
  }

  private handleClick(event) {
    if (event.target.getAttribute('aria-disabled') === 'true') {
      event.preventDefault()
      return
    }

    this.checked = event.target.checked
  }

  render() {
    return (
      <Host
        class={getClassNames([
          'ld-toggle',
          this.size === 'lg' && 'ld-toggle--lg',
          this.hasIcons && 'ld-toggle--with-icons',
        ])}
      >
        <input
          aria-disabled={this.ariaDisabled}
          checked={this.checked}
          disabled={this.disabled}
          onBlur={this.handleBlur.bind(this)}
          onClick={this.handleClick.bind(this)}
          onFocus={this.handleFocus.bind(this)}
          required={this.required}
          type="checkbox"
        />
        <span class="ld-toggle__knob" />
        {this.hasIcons && (
          <div class="ld-toggle__icon-start">
            <slot name="icon-start" />
          </div>
        )}
        {this.hasIcons && (
          <div class="ld-toggle__icon-end">
            <slot name="icon-end" />
          </div>
        )}
      </Host>
    )
  }
}
