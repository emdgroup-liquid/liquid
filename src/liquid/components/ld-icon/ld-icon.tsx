import { Build, Component, Host, h, Prop, Watch, Element } from '@stencil/core'
import { getClassNames } from '../../utils/getClassNames'
import { fetchIcon } from './fetchIcon'

/**
 * @slot - (optional) Custom SVG icon (only valid without name prop).
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part icon - Actual SVG element
 */
@Component({
  assetsDirs: ['assets'],
  tag: 'ld-icon',
  styleUrl: 'ld-icon.css',
  shadow: true,
})
export class LdIcon {
  @Element() element: HTMLElement

  /** The icon name. */
  @Prop() name: string = null

  /** Size of the icon. */
  @Prop() size?: 'sm' | 'lg'

  @Watch('name')
  private async loadIconPathData(): Promise<void> {
    if ((!Build.isBrowser && !Build.isTesting) || !this.name) {
      return
    }

    const div = document.createElement('div')
    const iconString = await fetchIcon(this.name)

    div.innerHTML = iconString.replace(
      '<svg',
      '<svg class="ld-icon__svg" part="icon"'
    )
    Array.from(this.element.shadowRoot.children).forEach((child) => {
      if (child.tagName !== 'STYLE') {
        this.element.shadowRoot.removeChild(child)
      }
    })
    this.element.shadowRoot.appendChild(div.firstChild)
  }

  async componentWillLoad(): Promise<void> {
    this.loadIconPathData()
  }

  render() {
    return (
      <Host
        class={getClassNames(['ld-icon', this.size && `ld-icon--${this.size}`])}
        role="presentation"
      >
        {!this.name && <slot></slot>}
      </Host>
    )
  }
}
