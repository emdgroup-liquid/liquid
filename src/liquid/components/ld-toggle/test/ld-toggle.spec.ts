import { newSpecPage } from '@stencil/core/testing'
import { LdIcon } from '../../ld-icon/ld-icon'
import { LdToggle } from '../ld-toggle'

describe('ld-toggle', () => {
  it('renders default', async () => {
    const page = await newSpecPage({
      components: [LdToggle],
      html: `<ld-toggle></ld-toggle>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders disabled state', async () => {
    const page = await newSpecPage({
      components: [LdToggle],
      html: `<ld-toggle disabled></ld-toggle>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders large mode', async () => {
    const page = await newSpecPage({
      components: [LdToggle],
      html: `<ld-toggle size="lg"></ld-toggle>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders with icons', async () => {
    const page = await newSpecPage({
      components: [LdIcon, LdToggle],
      html: `<ld-toggle>
        <ld-icon name="placeholder" size="sm" slot="icon-start"></ld-icon>
        <ld-icon name="placeholder" size="sm" slot="icon-end"></ld-icon>
      </ld-toggle>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('updates checked prop on checked value change', async () => {
    const page = await newSpecPage({
      components: [LdToggle],
      html: `<ld-toggle></ld-toggle>`,
    })
    const ldToggle = page.root
    expect(ldToggle.checked).toBe(undefined)

    const input = ldToggle.shadowRoot.querySelector('input')
    expect(input).toHaveProperty('checked', false)

    ldToggle.dispatchEvent(new Event('click', { bubbles: true }))
    await page.waitForChanges()
    expect(input).toHaveProperty('checked', true)
    expect(ldToggle.checked).toBe(true)

    expect(page.root).toMatchSnapshot()
  })

  it('emits focus and blur event', async () => {
    const page = await newSpecPage({
      components: [LdToggle],
      html: `<ld-toggle></ld-toggle>`,
    })
    const ldToggle = page.root
    const input = ldToggle.shadowRoot.querySelector('input')

    const handlers = {
      onFocus() {
        return
      },
      onBlur() {
        return
      },
    }

    const spyFocus = jest.spyOn(handlers, 'onFocus')
    ldToggle.addEventListener('focus', handlers.onFocus)
    input.dispatchEvent(new Event('focus'))
    jest.advanceTimersByTime(0)
    expect(spyFocus).toHaveBeenCalled()

    const spyBlur = jest.spyOn(handlers, 'onBlur')
    ldToggle.addEventListener('blur', handlers.onBlur)
    input.dispatchEvent(new Event('blur'))
    jest.advanceTimersByTime(0)
    expect(spyBlur).toHaveBeenCalled()
  })

  it('prevents input value changes with an aria-disabled attribute', async () => {
    const page = await newSpecPage({
      components: [LdToggle],
      html: `<ld-toggle aria-disabled="true"></ld-toggle>`,
    })
    const ldToggle = page.root

    expect(ldToggle.checked).toBe(undefined)

    const input = ldToggle.shadowRoot.querySelector('input')

    ldToggle.dispatchEvent(new Event('click'))
    await page.waitForChanges()

    expect(input).toHaveProperty('checked', false)
    expect(ldToggle.checked).toBe(undefined)
    expect(page.root).toMatchSnapshot()
  })

  it('emits input event on click', async () => {
    const page = await newSpecPage({
      components: [LdToggle],
      html: `<ld-toggle />`,
    })
    const ldCheckbox = page.root

    const spyInput = jest.fn()
    ldCheckbox.addEventListener('input', spyInput)
    ldCheckbox.click()

    expect(spyInput).toHaveBeenCalled()
  })

  it('allows to set inner focus', async () => {
    const { root } = await newSpecPage({
      components: [LdToggle],
      html: `<ld-toggle />`,
    })
    const input = root.shadowRoot.querySelector('input')

    input.focus = jest.fn()
    await root.focusInner()

    expect(input.focus).toHaveBeenCalled()
  })

  it('creates hidden input field, if inside a form', async () => {
    const { root } = await newSpecPage({
      components: [LdToggle],
      html: `<form><ld-toggle name="example" /></form>`,
    })
    expect(root).toMatchSnapshot()
  })

  it('sets initial state on hidden input', async () => {
    const { root } = await newSpecPage({
      components: [LdToggle],
      html: `<form><ld-toggle name="example" checked /></form>`,
    })
    expect(root.querySelector('input')).toHaveProperty('name', 'example')
  })

  it('updates hidden input field', async () => {
    const { root, waitForChanges } = await newSpecPage({
      components: [LdToggle],
      html: `<form><ld-toggle name="example" /></form>`,
    })
    const ldToggle = root
    await waitForChanges()

    expect(ldToggle.querySelector('input')).toHaveProperty('name', 'example')
    expect(ldToggle.querySelector('input')).toHaveProperty('checked', false)

    ldToggle.setAttribute('name', 'test')
    await waitForChanges()

    expect(ldToggle.querySelector('input')).toHaveProperty('name', 'test')

    ldToggle.removeAttribute('name')
    await waitForChanges()

    expect(ldToggle.querySelector('input')).toEqual(null)

    ldToggle.setAttribute('name', 'test')
    await waitForChanges()

    expect(ldToggle.querySelector('input')).toHaveProperty('name', 'test')

    ldToggle.dispatchEvent(new Event('click'))
    await waitForChanges()

    expect(ldToggle.querySelector('input')).toHaveProperty('checked', true)

    ldToggle.setAttribute('value', 'test')
    await waitForChanges()

    expect(ldToggle.querySelector('input')).toHaveProperty('value', 'test')

    ldToggle.removeAttribute('value')
    await waitForChanges()

    expect(ldToggle.querySelector('input').getAttribute('value')).toEqual(null)
  })

  it('uses hidden input field with referenced form', async () => {
    const { root, waitForChanges } = await newSpecPage({
      components: [LdToggle],
      html: '<ld-toggle name="example" form="yolo" />',
    })
    const ldToggle = root
    await waitForChanges()

    expect(ldToggle.querySelector('input')).toHaveProperty('name', 'example')
  })
})
