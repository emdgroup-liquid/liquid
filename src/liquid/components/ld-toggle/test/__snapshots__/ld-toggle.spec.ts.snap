// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-toggle prevents input value changes with an aria-disabled attribute 1`] = `
<ld-toggle aria-disabled="true" class="ld-toggle">
  <mock:shadow-root>
    <input aria-disabled="true" part="input focusable" type="checkbox">
    <span class="ld-toggle__knob" part="knob"></span>
  </mock:shadow-root>
</ld-toggle>
`;

exports[`ld-toggle renders default 1`] = `
<ld-toggle class="ld-toggle">
  <mock:shadow-root>
    <input part="input focusable" type="checkbox">
    <span class="ld-toggle__knob" part="knob"></span>
  </mock:shadow-root>
</ld-toggle>
`;

exports[`ld-toggle renders disabled state 1`] = `
<ld-toggle class="ld-toggle" disabled="">
  <mock:shadow-root>
    <input disabled="" part="input focusable" type="checkbox">
    <span class="ld-toggle__knob" part="knob"></span>
  </mock:shadow-root>
</ld-toggle>
`;

exports[`ld-toggle renders large mode 1`] = `
<ld-toggle class="ld-toggle ld-toggle--lg" size="lg">
  <mock:shadow-root>
    <input part="input focusable" type="checkbox">
    <span class="ld-toggle__knob" part="knob"></span>
  </mock:shadow-root>
</ld-toggle>
`;

exports[`ld-toggle renders with icons 1`] = `
<ld-toggle class="ld-toggle ld-toggle--with-icons">
  <mock:shadow-root>
    <input part="input focusable" type="checkbox">
    <span class="ld-toggle__knob" part="knob"></span>
    <div class="ld-toggle__icon-start" part="icon-wrapper icon-wrapper-start">
      <slot name="icon-start"></slot>
    </div>
    <div class="ld-toggle__icon-end" part="icon-wrapper icon-wrapper-end">
      <slot name="icon-end"></slot>
    </div>
  </mock:shadow-root>
  <ld-icon name="placeholder" size="sm" slot="icon-start">
    <!---->
    <span class="ld-icon ld-icon--sm" role="presentation">
      Not Found
    </span>
  </ld-icon>
  <ld-icon name="placeholder" size="sm" slot="icon-end">
    <!---->
    <span class="ld-icon ld-icon--sm" role="presentation">
      Not Found
    </span>
  </ld-icon>
</ld-toggle>
`;

exports[`ld-toggle updates checked prop on checked value change 1`] = `
<ld-toggle checked="" class="ld-toggle">
  <mock:shadow-root>
    <input checked="" part="input focusable" type="checkbox">
    <span class="ld-toggle__knob" part="knob"></span>
  </mock:shadow-root>
</ld-toggle>
`;
